### ISSUE TITLE
unnecessary_safety_comment doesn't understand assignment without `let`

### ISSUE THREAD

**Issue by nazar-pc** [rust-lang/rust-clippy#17039]
### Summary

I'm fairly confident the false-positive is caused by less popular assignment. Here is the original code:
https://github.com/nazar-pc/abundance/blob/8bce569d34d3a2a93fab3b30716ac6ea773cea66/crates/shared/ab-merkle-tree/src/balanced.rs#L159-L161

Changing the code to this resolves false-positive:
```rust
// SAFETY: The size of the tree is statically known to match the number of leaves and
// levels of hashes
let (parent_hashes_tmp, tree_hashes_tmp) =
    unsafe { tree_hashes.split_at_mut_unchecked(num_pairs) };
parent_hashes = parent_hashes_tmp;
tree_hashes = tree_hashes_tmp;
```

### Lint Name

unnecessary_safety_comment

### Reproducer

I tried this code:

```rust
// SAFETY: The size of the tree is statically known to match the number of leaves and
// levels of hashes
(parent_hashes, tree_hashes) = unsafe { tree_hashes.split_at_mut_unchecked(num_pairs) };
```

I saw this happen:

```
warning: statement has unnecessary safety comment
   --> crates/shared/ab-merkle-tree/src/balanced.rs:161:13
    |
161 |             (parent_hashes, tree_hashes) = unsafe { tree_hashes.split_at_mut_unchecked(num_pairs) };
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
help: consider removing the safety comment
   --> crates/shared/ab-merkle-tree/src/balanced.rs:159:16
    |
159 |             // SAFETY: The size of the tree is statically known to match the number of leaves and
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_safety_comment
    = note: `-W clippy::unnecessary-safety-comment` implied by `-W clippy::restriction`
    = help: to override `-W clippy::restriction` add `#[allow(clippy::unnecessary_safety_comment)]`
```

I expected to see this happen:
Safety comment is correctly applied to `unsafe` block and should not trigger this lint.

### Version

```text
rustc 1.97.0-nightly (20de910db 2026-05-02)
binary: rustc
commit-hash: 20de910db49d3476ccf49ea79a4b22e2b5dface0
commit-date: 2026-05-02
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.4
```

**Comment by samueltardieu** [comment]
Would you be able to come up with a minimal reproducer?

**Comment by Gri-ffin** [comment]
minimal
```rust
#![warn(clippy::unnecessary_safety_comment)]

fn main() {
    let mut data = [1u8, 2, 3, 4];
    let mut slice = data.as_mut_slice();

    let parent;
    // SAFETY: length is within bounds
    (parent, slice) = unsafe { slice.split_at_mut_unchecked(slice.len() / 2) };

    let _ = (parent, slice);
}
```

**Comment by samueltardieu** [comment]
Interesting. Clippy gets confused by the desugaring of the compound assign statement, which introduces a (safe) block around the expansion. The safety comment is then attributed to this block.

**Comment by samueltardieu** [comment]
> [@rustbot](https://github.com/rustbot) claim

Sorry I missed the claim, I implemented a fix right when I posted the message and just uploaded it when I came back from dinner after RustWeek.

I'll reassign the issue to myself.

**Comment by Gri-ffin** [comment]
no worries
